sql/hive/src/main/java/org/apache/hadoop/hive/ql/io/orc/VectorizedSparkOrcNewRecordReader.java

```diff
@@ -0,0 +1,318 @@
+import org.apache.hadoop.hive.ql.exec.vector.LongColumnVector;
+import org.apache.hadoop.hive.ql.exec.vector.VectorizedRowBatch;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.apache.hadoop.mapreduce.lib.input.FileSplit;
+
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.util.ArrayData;
+import org.apache.spark.sql.catalyst.util.MapData;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.Decimal;
+import org.apache.spark.unsafe.types.CalendarInterval;
+import org.apache.spark.unsafe.types.UTF8String;
+
+/**
+ * A RecordReader that returns InternalRow for Spark SQL execution.
+ * This reader uses an internal reader that returns Hive's VectorizedRowBatch. An adapter
+ * class is used to return internal row by directly accessing data in column vectors.
+ */
+public class VectorizedSparkOrcNewRecordReader
+    extends org.apache.hadoop.mapreduce.RecordReader<NullWritable, InternalRow> {
+  private final org.apache.hadoop.mapred.RecordReader<NullWritable, VectorizedRowBatch> reader;
+  private final int numColumns;
+  private VectorizedRowBatch internalValue;
+  private float progress = 0.0f;
+  private List<Integer> columnIDs;
+
+  private long numRowsOfBatch = 0;
+  private int indexOfRow = 0;
+
+  private final Row row;
+
+  public VectorizedSparkOrcNewRecordReader(
+      Reader file,
+      JobConf conf,
+      FileSplit fileSplit,
+      List<Integer> columnIDs) throws IOException {
+    List<OrcProto.Type> types = file.getTypes();
+    numColumns = (types.size() == 0) ? 0 : types.get(0).getSubtypesCount();
+    this.reader = new SparkVectorizedOrcRecordReader(file, conf,
+      new org.apache.hadoop.mapred.FileSplit(fileSplit));
+
+    this.columnIDs = new ArrayList<>(columnIDs);
+    this.internalValue = this.reader.createValue();
+    this.progress = reader.getProgress();
+    this.row = new Row(this.internalValue.cols, this.columnIDs);
+  }
+
+  @Override
+  public void close() throws IOException {
+    reader.close();
+  }
+
+  @Override
+  public NullWritable getCurrentKey() throws IOException,
+      InterruptedException {
+    return NullWritable.get();
+  }
+
+  @Override
+  public InternalRow getCurrentValue() throws IOException,
+      InterruptedException {
+    if (indexOfRow >= numRowsOfBatch) {
+      return null;
+    }
+    row.rowId = indexOfRow;
+    indexOfRow++;
+
+    return row;
+  }
+
+  @Override
+  public float getProgress() throws IOException, InterruptedException {
+    return progress;
+  }
+
+  @Override
+  public void initialize(InputSplit split, TaskAttemptContext context)
+      throws IOException, InterruptedException {
+  }
+
+  @Override
+  public boolean nextKeyValue() throws IOException, InterruptedException {
+    if (indexOfRow == numRowsOfBatch && progress < 1.0f) {
```

Comment:
We can't rely on `progress` to indicate the last row -- `< 1.0f` can evaluate to true even before the last row because of imprecision. Anyway, this part of the check is redundant, `reader.next` on the next line will return false when no more rows remain.